### PR TITLE
perf(algorithm): compute dilated convs as phase convs where faster

### PR DESCRIPTION
**Pull request by YourSandwich**
MIOpen has no working dilated 1D convolution kernel on AMD gfx1100, and the ResBlock
dilations are where RVC training spends most of its time. Undilated `conv1d` on that card
hits 18.7 TFLOP/s. The identical FLOPs at dilation 3 run at 0.30 TFLOP/s.

Measured with bf16, batch 8, `C512 L2048 k3`:

| dilation | time | effective |
|---|---|---|
| 1 | 1.379 ms | 18.7 TFLOP/s |
| 3 | 85.140 ms | 0.30 TFLOP/s |
| 5 | 85.264 ms | 0.30 TFLOP/s |

That is a 62x penalty for the same arithmetic. It also explains the card sitting at "100% busy"
while drawing 203-269 W of a 402 W cap: it is stalled in that kernel.

## What this changes

A dilation-`D` convolution is exactly `D` independent undilated convolutions over the strided
phases `x[..., i::D]`. This adds that as an alternative path.

It does not force it. On the first dilated convolution of a process both paths are timed once and
the winner is kept for the rest of the run, with the native kernel keeping ties. A probe costs a few
milliseconds against a run of hours. That way this helps cards where MIOpen is bad and stays out of
the way on cards where it is not, without hardcoding a vendor, an architecture or a ROCm version.

The helper sits in `commons.py`, which is a leaf module, so every dilated convolution in the tree
can reach it without an import cycle. Four call sites are routed through it, one line each:

| file | site | dilations |
|---|---|---|
| `residuals.py` | `ResBlock.convs1` | `(1, 3, 5)` |
| `generators/hifigan_mrf.py` | `MRFLayer.conv1` | per-layer |
| `generators/refinegan.py` | `ResBlock.convs1` | `(1, 3, 5)` |
| `modules.py` | `WaveNet.in_layers` | `dilation_rate ** i` |

So HiFi-GAN, MRF HiFi-GAN and RefineGAN all benefit, not only the default vocoder. The second
convolution in each block is dilation 1 and short-circuits immediately. `WaveNet` ships with
`dilation_rate=1` in the current configs, so that path is inert today and correct if it ever is not.

## Numbers

The convolution itself is 22-33x faster depending on shape. End to end on RVC training, batch 8,
2790 clips, measured over steps 40-90 after warmup:

| vocoder | before | after | |
|---|---|---|---|
| HiFi-GAN (NSF) | 2.260 s/it, epoch 13:08 | **0.860 s/it, epoch 5:00** | **2.63x** |
| MRF HiFi-GAN | not measured | 0.860 s/it, epoch 5:00 | |
| RefineGAN | not measured | 1.100 s/it, epoch 6:23 | |

2.63x on HiFi-GAN takes 300 epochs from 66.7 h to 25.4 h, so 41 hours off one run. Those two
figures also put the dilated convolutions at roughly 64% of the step, which matches where the time
was going.

## Correctness

The output is equivalent, not approximate. I checked every routed block against the native
formulation, forward and backward:

| block | forward | backward mean | cosine |
|---|---|---|---|
| `residuals.ResBlock` | 1.80e-07 | 2.27e-08 | 1.000000000 |
| `hifigan_mrf.MRFLayer` | 1.46e-07 | 4.54e-07 | 0.999999881 |
| `refinegan.ResBlock` | 2.89e-07 | 8.98e-07 | 1.000000000 |
| `modules.WaveNet` (d=2) | 2.90e-07 | 6.74e-08 | 1.000000000 |

A single convolution on its own comes out at `3e-07` on the backward pass.

The *max* backward error reads `1e-3`. That sits in about 0.01% of elements: `leaky_relu` has a
discontinuous derivative at zero, so a `2e-7` forward difference flips the branch for activations
that close to the kink and changes those gradients by 10x. Mean error and cosine similarity are the
meaningful measures. These models train in bf16 anyway, whose ~4e-3 relative precision cannot
represent the difference.

## Environment

Measured everything above on:

| | |
|---|---|
| GPU | Radeon RX 7900 XTX, `gfx1100`, 24 GB |
| ROCm | 7.2.4 |
| OS | Arch Linux, kernel 7.2.3 |
| Python | 3.12.13 |
| torch | 2.11.0+rocm7.2 (HIP 7.2.26015) |
| torchaudio | 2.11.0+rocm7.2 |
| triton | `triton_rocm` 3.6.0 |

The ROCm stack came from the PyTorch ROCm wheel index, not from a system ROCm install:

```bash
uv venv --python 3.12 .venv
uv pip install --python .venv/bin/python \
  --index-url https://download.pytorch.org/whl/rocm7.2 torch torchaudio
uv pip install --python .venv/bin/python -r requirements.txt   # torch lines removed
```

Two things worth knowing if you reproduce it. The wheel index is named by ROCm *minor*
version (`rocm7.2`) while AMD's own wheel repo uses the patch version (`rocm-rel-7.2.4`); they are
not interchangeable. And `triton-rocm` comes in with the torch wheel automatically. Installing
plain `triton` from PyPI pulls the CUDA build and breaks compiled kernels.

MIOpen ships inside the torch wheel, so the missing dilated kernel is a property of that wheel
rather than of the system ROCm install.

## Scope and risk

Nothing is gated on vendor or version, so this does not assume every ROCm build or every AMD
architecture shares the gfx1100 problem, and it does not assume cuDNN is fine either. The
measurement decides per machine. Where the native kernel wins, the only cost is one probe at
startup.

I measured this on gfx1100 / ROCm 7.2.4 / torch 2.11.0, where the probe picks the rewrite. I have no
CDNA card or NVIDIA card to confirm the other branch is taken there, so that path is reasoned
rather than observed.

### CODE REVIEW

**Comment by AznamirWoW** [comment]
Perhaps you can just do the performance test at the start once?

**Comment by YourSandwich** [comment]
@AznamirWoW Thanks for looking at it.
 The reason it is lazy rather than at import or `__init__` is the device. The model is built on CPU
and moved with `.to(device)` afterwards, so a probe at construction would benchmark CPU and decide
nothing useful about the GPU it will actually run on. The first forward is the first point where the device is known, and it also means the probe times the real tensor shape.

If you would rather have it explicit, I could add a small `set_dilated_path(use_phases)` so
`train.py` and the inference pipeline can decide once at startup and skip the probe entirely.

**Comment by AznamirWoW** [comment]
I mean you can add it rvc/lib/zluda.py
check whether the GPU name has 'AMD' in it, then do the speed test with a random tensor, etc.

**Comment by YourSandwich** [comment]
True, did not think about that.
Thanks for maintaining this project. <3

**Comment by AznamirWoW** [comment]
I need to test this with my old 6700XT. I believe TheROCK falls back to GEMM implementation while doing the normal dilated convolution.

Please re-test the native dilated conv1d with the following (you need to install rocm_sdk_devel).

# -----------------------------------------------------------------------------
# MIOpen Environment Setup (must precede the Torch import)
# -----------------------------------------------------------------------------
os.environ["MIOPEN_FIND_MODE"] = "2"
os.environ["MIOPEN_FIND_ENFORCE"] = "2"
# Enforce mode 1 does not necessarily pick the fastest kernel, 
# Enforce mode 2 is the only one that guarantees the fastest kernel is chosen.
os.environ["MIOPEN_SYSTEM_DB_PATH"] = "env\\Lib\\site-packages\\_rocm_sdk_devel\\bin\\"
os.environ["ROCBLAS_TENSILE_LIBPATH"] = "env\\Lib\\site-packages\\_rocm_sdk_libraries\\bin\\rocblas\\library"
os.environ["ROCBLAS_LAYER"] = "0"
os.environ["PYTORCH_ROCM_USE_ROCBLAS"] = "0"
os.environ["MIOPEN_ENABLE_LOGGING"] = "1"
os.environ["MIOPEN_ENABLE_LOGGING_ELAPSED_TIME"] = "1"
os.environ["MIOPEN_LOG_LEVEL"] = "5"
os.environ["MIOPEN_DEBUG_ENABLE"] = "0"
os.environ["MIOPEN_CONVOLUTION_MAX_WORKSPACE"] = "1073741824"  # 1 GiB

**Comment by YourSandwich** [comment]
You are right about GEMM, MIOpen picks `GemmFwdRest`. But no find mode changes it.

gfx1100, torch 2.11.0+rocm7.2, bf16, `N8 C512 L2048 k3`, dilation 3:

| config | native | phases |
|---|---|---|
| default | 85.56 ms | 1.77 ms |
| `FIND_ENFORCE=3` (Search) | 85.52 ms | 1.72 ms |
| `FIND_MODE=1` (Normal) | 85.53 ms | 1.76 ms |
| `FIND_MODE=1` + `ENFORCE=3` | 85.37 ms | 1.77 ms |

The solver log says why. Full find, nothing skipped: 68 solvers considered, 66 `Not applicable`, two survive.

```
ConvDirectNaiveConvFwd: Success.
GemmFwdRest: Success.
```

Naive and GEMM. There is no dilated kernel in the applicable set, so a search has nothing to find. The heuristic drops out as well, since torch hands this over as a 2D conv with a 1x3 filter: `TunaNet Inapplicable: Filters must be square (fil_h == fil_w)`.

Your two values are not the exhaustive ones either. `MIOPEN_FIND_MODE=2` is `Fast`, `Normal` is `1`. `MIOPEN_FIND_ENFORCE=2` is `DbUpdate`, `Search` is `3`. I ran your combination anyway, 85.44 ms.

**Comment by AznamirWoW** [comment]
I ran a test on 6700xt. Without setting MIOPEN_FIND_MODE the native method is slower. With find mode 2 native is usually faster than phases.

**Review by AznamirWoW — APPROVED** [review]

**Comment by YourSandwich** [comment]
I'm onto something

**Comment by YourSandwich** [comment]
It is `MIOPEN_FIND_MODE=2` on its own: 85.40 ms -> 1.35 ms on gfx1100. Adding `FIND_ENFORCE=2` puts it back to 85.48, which is why my last comment said no change.

So native beats this patch, 1.35 ms against 1.77 ms, and dilation 1 is unchanged either way.

Did you have `FIND_ENFORCE=2` set on the 6700XT run, or only the mode? Curious whether it cancels it there too.

**Comment by YourSandwich** [comment]
It might be worth documenting or keeping this patch as a fallback.

**Comment by YourSandwich** [comment]
Or `zluda.py` could do this, but who knows to which cards this applies :<
```python
if torch.cuda.is_available() and "AMD" in torch.cuda.get_device_name():
      os.environ.setdefault("MIOPEN_FIND_MODE", "2")
```

**Comment by YourSandwich** [comment]
Also, Zluda is not being imported into `infer.py`. Gradio inference gets it, `core.py infer` does not.

With `MIOPEN_FIND_MODE=2` a 10 s conversion here goes  31.2 s -> 6.2 s, so the CLI path is not covered by this patch.

**Comment by YourSandwich** [comment]
Pushed both, so this is now the `setdefault` plus the `infer.py` import instead of the conv1d rewrite.

`setdefault` keeps it overridable and it only fires on AMD, so I dropped the timing probe. gfx1100 and your 6700XT both get faster and dilation 1 is unchanged here. If a card turns up where mode 2 does not help, that is a reason to bring the probe back rather than the rewrite.

Title and description still describe the old approach, happy to rewrite them.

**Comment by AznamirWoW** [comment]
> Did you have `FIND_ENFORCE=2` set on the 6700XT run, or only the mode? Curious whether it cancels it there too.

on 6700xt the enforce flag does not anything related to dilated convolutions, it may affect some other things

**Comment by YourSandwich** [comment]
I have a bad, unfortunate finding regarding training with these variable. Let me sum that up

**Comment by YourSandwich** [comment]
`MIOPEN_FIND_MODE=2` is faster on an isolated convolution but much slower in real training, so the rewrite stays.

Same ~160 steps of the actual trainer each time, gfx1100, batch 4. Rewrite alone 2.68 it/s, rewrite plus the variable 1.01, variable alone 0.87. They do not stack either. Setting the variable on top of the rewrite costs 2.7x. The rewrite only replaces the dilated convs, everything else still goes through MIOpen, and Fast mode misses the find db on every new shape. Training runs a lot of shapes through bucketing plus two discriminators. My earlier benchmark ran one shape, forward only.

Inference does not need the variable either. A 10 s conversion is 31.2 s stock, 6.4 s with the rewrite alone, 6.3 s with both. The first run on a cold MIOpen cache reads 22.8 s, which is kernel compilation rather than steady state.

Restored the patch and kept the `infer.py` import.
